File: js/plugins/compat-oai/src/image.ts
/**
 * Copyright 2024 The Fire Company
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import type {
  GenerateRequest,
  GenerateResponseData,
  ModelReference,
} from 'genkit';
import { Message, modelRef, z } from 'genkit';
import { ModelAction, ModelInfo } from 'genkit/model';
import { model } from 'genkit/plugin';
import OpenAI from 'openai';
import type {
  ImageGenerateParams,
  ImagesResponse,
} from 'openai/resources/images.mjs';
import { PluginOptions } from './index.js';
import { maybeCreateRequestScopedOpenAIClient } from './utils.js';

export type ImageRequestBuilder = (
  req: GenerateRequest,
  params: ImageGenerateParams
) => void;

export const IMAGE_GENERATION_MODEL_INFO: ModelInfo = {
  supports: {
    media: false,
    output: ['media'],
    multiturn: false,
    systemRole: false,
    tools: false,
  },
};

export const ImageGenerationCommonConfigSchema = z.object({
  size: z.enum(['1024x1024', '1792x1024', '1024x1792']).optional(),
  style: z.enum(['vivid', 'natural']).optional(),
  user: z.string().optional(),
  n: z.number().int().min(1).max(10).default(1),
  quality: z.enum(['standard', 'hd']).optional(),
  response_format: z.enum(['b64_json', 'url']).default('b64_json').optional(),
});

function toImageGenerateParams(
  modelName: string,
  request: GenerateRequest,
  requestBuilder?: ImageRequestBuilder
): ImageGenerateParams {
  const {
    temperature,
    version: modelVersion,
    maxOutputTokens,
    stopSequences,
    topK,
    topP,
    response_format,
    ...restOfConfig
  } = (request.config as any) ?? {};

  let options: ImageGenerateParams = {
    model: modelVersion ?? modelName,
    prompt: new Message(request.messages[0]).text,
    response_format: response_format || 'b64_json',
  };
  if (requestBuilder) {
    requestBuilder(request, options);
  } else {
    options = { ...options, ...restOfConfig };
  }
  for (const k in options) {
    if (options[k] === undefined) {
      delete options[k];
    }
  }
  return options;
}

function toGenerateResponse(result: ImagesResponse): GenerateResponseData {
  const images = result.data;
  if (!images) {
    return { finishReason: 'stop' };
  } else {
    const content = (result.data ?? []).map((image) => ({
      media: {
        contentType: 'image/png',
        url: image.url || `data:image/png;base64,${image.b64_json}`,
      },
    }));
    return { message: { role: 'model', content }, raw: result };
  }
}

/**
 * Method to define a new Genkit Model that is compatible with Open AI
 * Images API. 
 *
 * These models are to be used to create images from a user prompt.
 *
 * @param params An object containing parameters for defining the OpenAI
 * image model.
 * @param params.ai The Genkit AI instance.
 * @param params.name The name of the model.
 * @param params.client The OpenAI client instance.
 * @param params.modelRef Optional reference to the model's configuration and
 * custom options.

 * @returns the created {@link ModelAction}
 */
export function defineCompatOpenAIImageModel<
  CustomOptions extends z.ZodTypeAny = z.ZodTypeAny,
>(params: {
  name: string;
  client: OpenAI;
  pluginOptions?: PluginOptions;
  modelRef?: ModelReference<CustomOptions>;
  requestBuilder?: ImageRequestBuilder;
}): ModelAction<CustomOptions> {
  const {
    name,
    client: defaultClient,
    pluginOptions,
    modelRef,
    requestBuilder,
  } = params;
  const modelName = name.substring(name.indexOf('/') + 1);

  return model(
    {
      name,
      ...modelRef?.info,
      configSchema: modelRef?.configSchema,
    },
    async (request, { abortSignal }) => {
      const client = maybeCreateRequestScopedOpenAIClient(
        pluginOptions,
        request,
        defaultClient
      );
      const result = await client.images.generate(
        toImageGenerateParams(modelName!, request, requestBuilder),
        { signal: abortSignal }
      );
      return toGenerateResponse(result);
    }
  );
}

/** Image generation ModelRef helper, with reasonable defaults for
 * OpenAI-compatible providers */
export function compatOaiImageModelRef<
  CustomOptions extends z.ZodTypeAny = z.ZodTypeAny,
>(params: {
  name: string;
  info?: ModelInfo;
  configSchema?: CustomOptions;
  config?: any;
  namespace?: string;
}) {
  const {
    name,
    info = IMAGE_GENERATION_MODEL_INFO,
    configSchema,
    config = undefined,
    namespace,
  } = params;
  return modelRef({
    name,
    configSchema:
      configSchema || (ImageGenerationCommonConfigSchema as z.ZodObject<any>),
    info,
    config,
    namespace,
  });
}

File: py/packages/genkit-openai/src/genkit_openai/models/image.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""OpenAI-compatible image generation model for Genkit.

Provides image generation capabilities via the OpenAI Images API,
supporting models like DALL-E 3 and GPT-Image-1.
"""

from __future__ import annotations

from typing import Any

from openai import APIStatusError, AsyncOpenAI
from openai.types.images_response import ImagesResponse

from genkit import (
    Media,
    MediaPart,
    Message,
    ModelInfo,
    ModelRequest,
    ModelResponse,
    Part,
    Role,
    Supports,
)
from genkit.model import FinishReason
from genkit.plugin_api import ActionRunContext
from genkit_openai.models.utils import _extract_text, extract_config_dict, reraise_openai_error

# GPT Image 1 has a different configuration surface from DALL-E models.
_GPT_IMAGE_1_CONFIG_SCHEMA: dict[str, Any] = {
    'type': 'object',
    'properties': {
        'size': {
            'type': 'string',
            'enum': ['1024x1024', '1536x1024', '1024x1536', 'auto'],
        },
        'style': {'type': 'string', 'enum': ['vivid', 'natural']},
        'user': {'type': 'string'},
        'n': {'type': 'integer', 'minimum': 1, 'maximum': 10, 'default': 1},
        'quality': {'type': 'string', 'enum': ['low', 'medium', 'high']},
        'background': {'type': 'string', 'enum': ['transparent', 'opaque', 'auto']},
        'moderation': {'type': 'string', 'enum': ['low', 'auto']},
        'output_compression': {'type': 'integer', 'minimum': 1, 'maximum': 100},
        'output_format': {'type': 'string', 'enum': ['png', 'jpeg', 'web']},
    },
}

# Supported image generation models with their metadata.
SUPPORTED_IMAGE_MODELS: dict[str, ModelInfo] = {
    'dall-e-3': ModelInfo(
        label='OpenAI - DALL-E 3',
        supports=Supports(
            media=False,
            output=['media'],
            multiturn=False,
            system_role=False,
            tools=False,
        ),
    ),
    'gpt-image-1': ModelInfo(
        label='OpenAI - GPT Image 1',
        config_schema=_GPT_IMAGE_1_CONFIG_SCHEMA,
        supports=Supports(
            media=False,
            output=['media'],
            multiturn=False,
            system_role=False,
            tools=False,
        ),
    ),
}


# Re-export _extract_text as _extract_prompt_text for backward compatibility.
_extract_prompt_text = _extract_text


def _to_image_generate_params(
    model_name: str,
    request: ModelRequest,
) -> dict[str, Any]:
    """Convert a ModelRequest into OpenAI image generation parameters.

    Extracts the text prompt and maps Genkit config options to OpenAI's
    image generation API parameters.

    Args:
        model_name: The OpenAI model name (e.g., 'dall-e-3').
        request: The Genkit generate request.

    Returns:
        A dictionary of parameters for client.images.generate().
    """
    prompt = _extract_prompt_text(request)
    config = extract_config_dict(request)

    has_snake_case_response_format = 'response_format' in config
    response_format = config.pop('response_format', None)
    camel_case_response_format = config.pop('responseFormat', None)
    if not has_snake_case_response_format:
        response_format = camel_case_response_format

    # Start with required params.
    effective_model = config.pop('version', None) or model_name
    params: dict[str, Any] = {
        'model': effective_model,
        'prompt': prompt,
    }

    # GPT Image 1 rejects the response_format parameter; its API always
    # returns base64 image data. DALL-E models retain the existing default.
    if 'gpt-image' not in effective_model.lower():
        params['response_format'] = response_format or 'b64_json'

    # Strip standard GenAI config keys that don't apply to image generation.
    for key in (
        'temperature',
        'max_output_tokens',
        'maxOutputTokens',
        'stop_sequences',
        'stopSequences',
        'top_k',
        'topK',
        'top_p',
        'topP',
        'api_key',
        'apiKey',
    ):
        config.pop(key, None)

    # Pass remaining config through (size, quality, style, n, etc.).
    params.update(config)

    # Remove None values.
    return {k: v for k, v in params.items() if v is not None}


def _to_generate_response(result: ImagesResponse) -> ModelResponse:
    """Convert an OpenAI ImagesResponse to a Genkit ModelResponse.

    Each generated image becomes a media part in the response message.

    Args:
        result: The OpenAI images.generate() response object.

    Returns:
        A ModelResponse with media parts for each generated image.
    """
    images = result.data
    if not images:
        return ModelResponse(
            message=Message(role=Role.MODEL, content=[]),
            finish_reason=FinishReason.STOP,
        )

    content: list[Part] = []
    for image in images:
        url = image.url
        if not url and image.b64_json:
            url = f'data:image/png;base64,{image.b64_json}'

        if url:
            content.append(Part(root=MediaPart(media=Media(content_type='image/png', url=url))))

    return ModelResponse(
        message=Message(role=Role.MODEL, content=content),
        finish_reason=FinishReason.STOP,
    )


class OpenAIImageModel:
    """Handles image generation via the OpenAI Images API.

    Args:
        model_name: The image model to use (e.g., 'dall-e-3').
        client: An async OpenAI client instance.
    """

    def __init__(self, model_name: str, client: AsyncOpenAI) -> None:
        """Initialize the image model.

        Args:
            model_name: The image model to use (e.g., 'dall-e-3').
            client: An async OpenAI client instance.
        """
        self._model_name = model_name
        self._client = client

    @property
    def name(self) -> str:
        """The name of the image model."""
        return self._model_name

    async def generate(self, request: ModelRequest, ctx: ActionRunContext) -> ModelResponse:
        """Generate images from the request.

        Args:
            request: The generate request containing the text prompt.
            ctx: The action run context.

        Returns:
            A ModelResponse containing generated image media parts.
        """
        try:
            params = _to_image_generate_params(self._model_name, request)
            result = await self._client.images.generate(**params)
            return _to_generate_response(result)
        except (APIStatusError, ValueError) as e:
            reraise_openai_error(e)
